pkg/apis/validation/validation.go

```diff
@@ -114,53 +114,45 @@ func validateClassSpecTags(tags map[string]string, fldPath *field.Path) field.Er
 
 // validateSecret validates that the secret contain data to authenticate with an Openstack provider.
 func validateSecret(secret *corev1.Secret) field.ErrorList {
-	var (
-		ok, ok2 bool
-		allErrs = field.ErrorList{}
-	)
+	allErrs := field.ErrorList{}
 
 	root := field.NewPath("data")
 	data := secret.Data
-	if b, ok := data[OpenStackAuthURL]; !ok || isEmptyStringByteSlice(b) {
+	if isEmptyStringByteSlice(data[OpenStackAuthURL]) {
 		allErrs = append(allErrs, field.Required(root.Key(OpenStackAuthURL), fmt.Sprintf("%s is required", OpenStackAuthURL)))
 	}
-	if _, ok := data[OpenStackApplicationCredentialID]; !ok {
-		if b, ok := data[OpenStackUsername]; !ok || isEmptyStringByteSlice(b) {
+	if isEmptyStringByteSlice(data[OpenStackApplicationCredentialSecret]) {
+		if isEmptyStringByteSlice(data[OpenStackUsername]) {
 			allErrs = append(allErrs, field.Required(root.Key(OpenStackUsername), fmt.Sprintf("%s is required", OpenStackUsername)))
 		}
-		if b, ok := data[OpenStackPassword]; !ok || isEmptyStringByteSlice(b) {
+		if isEmptyStringByteSlice(data[OpenStackPassword]) {
 			allErrs = append(allErrs, field.Required(root.Key(OpenStackPassword), fmt.Sprintf("%s is required", OpenStackPassword)))
 		}
 	} else {
-		if b, ok := data[OpenStackApplicationCredentialID]; !ok || isEmptyStringByteSlice(b) {
-			allErrs = append(allErrs, field.Required(root.Key(OpenStackApplicationCredentialID), fmt.Sprintf("%s is required", OpenStackApplicationCredentialID)))
+		if isEmptyStringByteSlice(data[OpenStackApplicationCredentialID]) &&
+			(isEmptyStringByteSlice(data[OpenStackApplicationCredentialName]) || isEmptyStringByteSlice(data[OpenStackUsername])) {
+			allErrs = append(allErrs, field.Required(root.Key(OpenStackApplicationCredentialID), fmt.Sprintf("%s or %s and %s are required if %s present", OpenStackApplicationCredentialID, OpenStackApplicationCredentialName, OpenStackUsername, OpenStackApplicationCredentialSecret)))
+			if isEmptyStringByteSlice(data[OpenStackApplicationCredentialName]) {
+				allErrs = append(allErrs, field.Required(root.Key(OpenStackApplicationCredentialName), fmt.Sprintf("%s or %s and %s are required if %s present", OpenStackApplicationCredentialID, OpenStackApplicationCredentialName, OpenStackUsername, OpenStackApplicationCredentialSecret)))
+			}
+			if isEmptyStringByteSlice(data[OpenStackUsername]) {
+				allErrs = append(allErrs, field.Required(root.Key(OpenStackUsername), fmt.Sprintf("%s or %s and %s are required if %s present", OpenStackApplicationCredentialID, OpenStackApplicationCredentialName, OpenStackUsername, OpenStackApplicationCredentialSecret)))
+			}
 		}
-		if b, ok := data[OpenStackApplicationCredentialSecret]; !ok || isEmptyStringByteSlice(b) {
+		if isEmptyStringByteSlice(data[OpenStackApplicationCredentialSecret]) {
```

Comment:
Don't we check that already in line 124?

Reply:
Right, missed in the last refactoring. Fixed.